sql: collated strings, phase three

eisenstatdavid opened the pull request:
cc @RaduBerinde @danhhz

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13609)
<!-- Reviewable:end -->

Commit:
util/encoding: improve panic formats for decoding errors

Commit:
sql/distsqlrun: do not call Finish on a nil span

RaduBerinde commented:
Review status: 0 of 19 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/server.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2ZlVCFnGuZq3Vy1TC:-Kd2ZlVCFnGuZq3Vy1TD:b-3f9zw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/23f7f691586508ba3e8b8a5e352d5a9a63164d98/pkg/sql/distsqlrun/server.go#L114)):*
> ```Go
> 	if err := f.setupFlow(ctx, &req.Flow); err != nil {
> 		log.Error(ctx, err)
> 		if sp != nil {
> ```

[nit] we usually use `tracing.FinishSpan(sp)` for this

---

*[pkg/sql/sqlbase/structured.proto, line 248 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_T_pkJ8UfB7Bxu3V:-Kd2_T_pkJ8UfB7Bxu3W:b-o470ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/structured.proto#L248)):*
> ```protobuf
>       (gogoproto.casttype) = "ColumnID"];
> 
>   // For secondary indexes, an ordered list of IDs for the collated string
> ```

I am wondering if we should make this more generic - there may be other things in the future that require a key + something else to store as value (a potential one is decimals, see #13384). Seems like the code using this could check the type of each column, making it possible to add more types like this later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 0 of 19 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/structured.proto, line 248 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_T_pkJ8UfB7Bxu3V:-Kd2ateNGVauCgtxtAA-:b-hv05ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/structured.proto#L248)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I am wondering if we should make this more generic - there may be other things in the future that require a key + something else to store as value (a potential one is decimals, see #13384). Seems like the code using this could check the type of each column, making it possible to add more types like this later.
</blockquote></details>

non-point geodata as well

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609)*
<!-- Sent from Reviewable.io -->

knz commented:
This is very good work. I am rather impressed.

Meanwhile it is also the case the 3rd commit introduced a rather big conceptual change in the way we should think about the SQL->KV mapping. I am uncomfortable that this change is merely listed in the commit message.

What I think would be best would be to create a (smallish) text file next to `design.md` in `docs/` where we talk about this mapping and the encoding of keys and values. In this file make a small bullet point list and explain how we encode all the types, both as keys and as values (I don't remember whether there's a difference). The largest entry is probably not longer than a few sentences with an example, talking about the encoding of collated strings. Also note somewhere where the code for encoding and decoding lives in the source tree.

Also the exercise of doing this may inspire either you or another reader to recognize some commonalities and/or future reuse opportunities and do something about it. (A softer version of what Radu was suggesting in his comment.)

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 16 of 16 files at r3.
Review status: 18 of 19 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/rowwriter.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2cCac3AAUquEMY3iV:-Kd2cCadK55n9XeUTNcw:b-6sgn0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/rowwriter.go#L109)):*
> ```Go
> 
> func (rh *rowHelper) isCollatedStringColumn(colID sqlbase.ColumnID) bool {
> 	if rh.colType == nil {
> ```

Huh why store all the columns in `colType`. Why not using a smaller map with only the collated string columns in there. Even a `map[ID]struct{}` would probably do.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 351 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2Zx3MFIAi7wN5xWlK:-Kd2Zx3N_A_y02_QYgEt:b-59hp52) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/rowfetcher.go#L351)):*
> ```Go
> 				return "", "", err
> 			}
> 			if idx, ok := rf.colIdxMap[id]; ok && rf.valNeededForCol[idx] {
> ```

Can you clarify here (perhaps in a comment) how it is possible that a column id is in `CollatedStringColumnIDs` but not in`colIdxMap`.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 481 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_5iKDLW6iuee_UcG:-Kd2_5iKDLW6iuee_UcH:b-m9ffx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/rowfetcher.go#L481)):*
> ```Go
> 			fmt.Fprintf(&rf.prettyValueBuf, "/%v", encValue.Datum)
> 		}
> 		if !rf.row[idx].IsUnset() && rf.cols[idx].Type.Kind != ColumnType_COLLATEDSTRING {
> ```

Please explain how/when this condition would be true if you weren't adding the check on the column type. (I don't understand it yet)

---

*[pkg/sql/sqlbase/structured.proto, line 248 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_pzJ7_fn5IwDwhLo:-Kd2_pzK6lA1TWZLhPQJ:byqabx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/structured.proto#L248)):*
> ```protobuf
>       (gogoproto.casttype) = "ColumnID"];
> 
>   // For secondary indexes, an ordered list of IDs for the collated string
> ```

This comment needs more words. I suggest: "CollatedStringColumnIDs contains the ordered list of IDs of columns that are listed in the index and have type collated string."

Also please clarify whether this includes only IDs from ColumnIDs, or both ColumnIDs and ExtraColumnIDs.

---

*[pkg/sql/testdata/collatedstring_uniqueindex2, line 29 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2bUiBT2rTdUDRKXb3:-Kd2bUiBT2rTdUDRKXb4:b-km50ux) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/testdata/collatedstring_uniqueindex2#L29)):*
> ```
> ----
> CREATE UNIQUE INDEX ON t (a, b);
> ```

Please provide a link in a comment here to the collating table for locale `de`. So that the reader of this test can check whether the test itself is correct.
Same for the other test files.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Kd2aTLd09cUcr74kQue:b-1hjika)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 16 of 16 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Addressed the comments except for the document. Will ping again when I have a draft.

---

Review status: 0 of 19 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/rowwriter.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2cCac3AAUquEMY3iV:-KdWkeWpQ3Hu3n-p855R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/rowwriter.go#L109)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Huh why store all the columns in `colType`. Why not using a smaller map with only the collated string columns in there. Even a `map[ID]struct{}` would probably do.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/server.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2ZlVCFnGuZq3Vy1TC:-KdWTfjyLOZMOXHDbSNW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/23f7f691586508ba3e8b8a5e352d5a9a63164d98/pkg/sql/distsqlrun/server.go#L114)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] we usually use `tracing.FinishSpan(sp)` for this
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 351 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2Zx3MFIAi7wN5xWlK:-KdWpf-eRKHbMockpE4f:b-c7g2s5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/rowfetcher.go#L351)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you clarify here (perhaps in a comment) how it is possible that a column id is in `CollatedStringColumnIDs` but not in`colIdxMap`.
</blockquote></details>

Collated string column that is not requested. I don't really understand the need to check `valNeededForCol` afterward, though, since I copied this condition from the extra columns code.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 481 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_5iKDLW6iuee_UcG:-KdWsFRM0lnGiy86fTyx:b-98v7b4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/rowfetcher.go#L481)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please explain how/when this condition would be true if you weren't adding the check on the column type. (I don't understand it yet)
</blockquote></details>

Done (and generalized the condition to reduce the number of places that collated strings are assumed).

---

*[pkg/sql/sqlbase/structured.proto, line 248 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_pzJ7_fn5IwDwhLo:-KdWnXz-qMj-c0qx6yrD:b6f5yhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/structured.proto#L248)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This comment needs more words. I suggest: "CollatedStringColumnIDs contains the ordered list of IDs of columns that are listed in the index and have type collated string."

Also please clarify whether this includes only IDs from ColumnIDs, or both ColumnIDs and ExtraColumnIDs.
</blockquote></details>

Done (yes, both -- I'll expand on my reasoning when I write the side doc).

---

*[pkg/sql/sqlbase/structured.proto, line 248 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_T_pkJ8UfB7Bxu3V:-KdWnZH84m3kCVox5y2s:b-hoqvd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/structured.proto#L248)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

non-point geodata as well
</blockquote></details>

Renamed to SplitKeyColumnIDs.

---

*[pkg/sql/testdata/collatedstring_uniqueindex2, line 29 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2bUiBT2rTdUDRKXb3:-KdWgPk-K8DSQslSipYa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/testdata/collatedstring_uniqueindex2#L29)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please provide a link in a comment here to the collating table for locale `de`. So that the reader of this test can check whether the test itself is correct.
Same for the other test files.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-KdWwBzza_uL-v1FO6T3:b-s202ru)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/sqlbase/structured.proto, line 248 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_T_pkJ8UfB7Bxu3V:-KdWx37YUxvw7FMiu2Aq:b-bbl0g9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/structured.proto#L248)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

Renamed to SplitKeyColumnIDs.
</blockquote></details>

Sorry to bikeshed, but "split key" is an overloaded term in our system..  Maybe `CompositeColumnIDs` (in that they have both key and value parts)? Maybe @danhhz or @knz have better suggestions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609)*
<!-- Sent from Reviewable.io -->

knz commented:
Ok now that I understand what the tests are supposed to do, I have to ask: does this implementation support unicode normalization properly? I am not sure because the tests currently use UTF-8 and the viewing tools (github, reviewable, my text editor) display the same whether the characters are normalized or not.

May I suggest to have at least a test which encodes the strings manually using the individual bytes in a byte array followed by a conversion to string, to expose the different UTF-8 encodings for the same character. Then check that it does the right thing wrt ordering and equality.

The rest LGTM. Curious about doc.

---

Reviewed 1 of 19 files at r5, 1 of 1 files at r6, 16 of 16 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/rowwriter.go, line 52 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KdWyzXvywT-vJjvjQ5R:-KdWyzXvywT-vJjvjQ5S:b-hc7oyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9c59d694179de620ddbf3db403ed2dd7b3acd1f/pkg/sql/rowwriter.go#L52)):*
> ```Go
> 	primaryIndexKeyPrefix []byte
> 	primaryIndexCols      map[sqlbase.ColumnID]struct{}
> 	splitKeyCols          map[sqlbase.ColumnID]struct{}
> ```

Same comment as Radu about the name "splitKey" here.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 351 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2Zx3MFIAi7wN5xWlK:-KdWzOUlUGkmKP6ZmR8g:b-kpei66) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/rowfetcher.go#L351)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

Collated string column that is not requested. I don't really understand the need to check `valNeededForCol` afterward, though, since I copied this condition from the extra columns code.
</blockquote></details>

What do you not understand? Also @RaduBerinde might be able to clarify.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 84 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KdWzEv768OoLCZEOLuc:-KdWzEv768OoLCZEOLud:b226wiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9c59d694179de620ddbf3db403ed2dd7b3acd1f/pkg/sql/sqlbase/rowfetcher.go#L84)):*
> ```Go
> 
> 	// IDs of split key columns.
> 	splitKeyCols map[ColumnID]struct{}
> ```

ditto about the name

---

*[pkg/sql/sqlbase/structured.proto, line 248 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_T_pkJ8UfB7Bxu3V:-KdWzu_58YyILlG329rD:b-ebwa3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/structured.proto#L248)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Sorry to bikeshed, but "split key" is an overloaded term in our system..  Maybe `CompositeColumnIDs` (in that they have both key and value parts)? Maybe @danhhz or @knz have better suggestions.
</blockquote></details>

In my view these are not "split" but indeed either "composite", "non-canonical", or "binary". I'm not good at names really. :smile_cat: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-KdXp-AUL7bTR1IBLbrh:b5n2qdw)*
<!-- Sent from Reviewable.io -->

Commit:
sql: collated strings, phase three

Collated strings can now be used as keys. This is accomplished by
encoding collated strings as the bytes encoding of their collation key
and then encoding the actual string value somewhere else. For primary
indexes (i.e., tables), collated string key parts are encoded as part of
the column family 0 value tuple. For secondary indexes, collated string
key parts are encoded at the end of the value.

See docs/encoding.md for a complete rundown on encoding.

One known cosmetic issue is that the error message for inserting a
duplicate key containing a collated string is terrible. I'll file an
issue shortly after this PR is merged.

Fixes #2473

Commit:
Revert "distsql: don't use collated strings in TestHashRouter"

This reverts commit ec476859e1b6d7cf2e5b7e28527edd89b59daed5.

eisenstatdavid commented:
Step 1 of the algorithm to derive the collation key is normalization, so this should just work. Nevertheless, I added a test.

Still working on the doc.

---

Review status: 1 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/rowwriter.go, line 52 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KdWyzXvywT-vJjvjQ5R:-Kdalju-FM7RSgc2F77t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9c59d694179de620ddbf3db403ed2dd7b3acd1f/pkg/sql/rowwriter.go#L52)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Same comment as Radu about the name "splitKey" here.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 351 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2Zx3MFIAi7wN5xWlK:-KdalrS9Bg29Jkt2r0AB:bfiuq3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/rowfetcher.go#L351)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

What do you not understand? Also @RaduBerinde might be able to clarify.
</blockquote></details>

Why non-needed columns would be assigned indexes.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 84 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KdWzEv768OoLCZEOLuc:-Kdall240HDQk4xURQmt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9c59d694179de620ddbf3db403ed2dd7b3acd1f/pkg/sql/sqlbase/rowfetcher.go#L84)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto about the name
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 248 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2_T_pkJ8UfB7Bxu3V:-Kdb59kQIl3d_MNxw0lv:b6cefa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/structured.proto#L248)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

In my view these are not "split" but indeed either "composite", "non-canonical", or "binary". I'm not good at names really. :smile_cat: 
</blockquote></details>

Changed to composite.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-KdbCTGErUrM15fpFmSg:b-ms3si4)*
<!-- Sent from Reviewable.io -->

knz commented:
Could you rebase this and resolve the conflicts so that I can complete the reivew (right now Reviewable is utterly confused)

mjibson commented:
I'd just like to mirror the comments about making sure this code can be used for other types. I have a PR out now (#13813) that is blocked on full support until we can separate decimal values and keys in indexes like you do here for collated strings.

eisenstatdavid commented:
Rebased. I'm going to try to reduce the amount of collated string specific code next.

---

Review status: 1 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Ke4SaUAIhP5tO0ScIgw:b-b1og1u)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Ready for more review.

Unless I'm missing something, we should now be able to do decimals by changing three pieces of code: 1) HasCompositeKeyEncoding 2) rowfetcher.go needs a new case for decoding composite columns in secondary indexes 3) table.go needs a new case for encoding composite columns in secondary indexes.

Matt, how are we planning to make the decimal encoding backward compatible?

---

Review status: 1 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Ke5XoQOPKrxtLpdEtFo:b-18rzzy)*
<!-- Sent from Reviewable.io -->

mjibson commented:
If the decoding falls back to using the value in the index if the composite value isn't present, that would work for supporting backwards compatibility. Is that doable with the current structure?

---

Review status: 1 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Ke5ZpVk53VUsOaBfc1X:b-bspmvh)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
I believe so -- existing IndexDescriptors won't name the decimal columns as composite columns, so the special decoding logic won't apply, and the fallback will be to just decode the key as usual.

---

Review status: 1 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Ke5_SmvLMx55iGQPEn8:ba6orui)*
<!-- Sent from Reviewable.io -->

knz commented:
Code LGTM.
Curious about the doc.

---

Reviewed 1 of 1 files at r10, 18 of 19 files at r11, 1 of 1 files at r12.
Review status: 21 of 22 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 351 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2Zx3MFIAi7wN5xWlK:-Ke_qG8XsSxWJSY8Uygv:b-nl8mrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/rowfetcher.go#L351)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

Why non-needed columns would be assigned indexes.
</blockquote></details>

My guess: that's because they may be part of an ordering specified by the query, while the values are not needed for the rendered columns.
For example:
```
create table t.kv(k primary key, v int, index vv(v));
explain(metadata) SELECT k FROM (select * from t.kv ORDER BY v);
+-------+--------+-------+-------------+-----------------+----------+
| Level |  Type  | Field | Description |     Columns     | Ordering |
+-------+--------+-------+-------------+-----------------+----------+
|     0 | render |       |             | (k)             |          |
|     1 | scan   |       |             | (k, v[omitted]) | +v       |
|     1 |        | table | kv@vv       |                 |          |
|     1 |        | spans | ALL         |                 |          |
+-------+--------+-------+-------------+-----------------+----------+
(4 rows)
```

see `v` gets omitted (`!valNeededForCol`) although it's the defining column for the index (`vv`) selected for the scan.

Do you think this explains accurately the situation?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Ke_rUffCkc8AWP-NRKB:bda23pu)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 19 files at r11.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Added the doc to the PR.

---

Review status: 21 of 23 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 351 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kd2Zx3MFIAi7wN5xWlK:-KeeXOMpdoh_mHsXZpiD:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff1cfdbdd4e7eecb891b1275545a26033155775/pkg/sql/sqlbase/rowfetcher.go#L351)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

My guess: that's because they may be part of an ordering specified by the query, while the values are not needed for the rendered columns.
For example:
```
create table t.kv(k primary key, v int, index vv(v));
explain(metadata) SELECT k FROM (select * from t.kv ORDER BY v);
+-------+--------+-------+-------------+-----------------+----------+
| Level |  Type  | Field | Description |     Columns     | Ordering |
+-------+--------+-------+-------------+-----------------+----------+
|     0 | render |       |             | (k)             |          |
|     1 | scan   |       |             | (k, v[omitted]) | +v       |
|     1 |        | table | kv@vv       |                 |          |
|     1 |        | spans | ALL         |                 |          |
+-------+--------+-------+-------------+-----------------+----------+
(4 rows)
```

see `v` gets omitted (`!valNeededForCol`) although it's the defining column for the index (`vv`) selected for the scan.

Do you think this explains accurately the situation?
</blockquote></details>

Yes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-KeeXQ7CblZMuEGe1Ac1:bc25844)*
<!-- Sent from Reviewable.io -->

knz commented:
@danhhz can you be second reviewer for the new doc `encoding.md` please. Thanks

---

Reviewed 2 of 2 files at r13.
Review status: 22 of 23 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[docs/encoding.md, line 9 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek0XNM3SNUovRZqg9u:-Kek0XNM3SNUovRZqg9v:bhfewl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L9)):*
> ```Markdown
> encoding methods. The primary author is David Eisenstat
> &lt;<eisen@cockroachlabs.com>&gt;.
> ```

There needs to be a (short) explanation here that this encoding tends to change over time, that we have "format versions" that reflect these changes, and that we try to keep it backward-compatible.

Also mention here which specific format version this doc is explaining. (I think 3? @danhhz please confirm)

---

*[docs/encoding.md, line 40 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek1DYORm2hoUC7WIh8:-Kek1DYPNG1xSgmuKCWu:bk3w27e) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L40)):*
> ```Markdown
> family ID is zero, indicating the special column family to which the
> primary-key columns belong. The purpose of this exception is to retain
> rows whose non–primary-key data is all null.
> ```

The word "sentinel key" must be mentioned and defined here.

Please provide three example KV pairs here: one with no null values, one with null values on only some columns of a family, and one with null values on all columns but that is a PK (= example sentinel key)

---

*[docs/encoding.md, line 50 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek1uXVo6t0qRZnsG96:-Kek1uXVo6t0qRZnsG97:b409kz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L50)):*
> ```Markdown
> 
> CRDB creates KV keys by encoding the fields that comprise the key and
> concatenating. This encoding has an inverse because no field code word
> ```

"The encoding of individual fields is chosen so that it is always possible to decode a concatenated key into its original components."

Also you need to define the term "field code word".

---

*[docs/encoding.md, line 60 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek2QQ9iweujRwyWsze:-Kek2QQ9iweujRwyWszf:b-qoahl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L60)):*
> ```Markdown
> and, for some types, the length of the code word and/or the encoding.
> For more details, see
> [pkg/util/encoding/encoding.go](https://github.com/cockroachdb/cockroach/blob/master/pkg/util/encoding/encoding.go).
> ```

I think it would be useful to provide a table with some of the most commonly used SQL types and their code word, as examples, here.

---

*[docs/encoding.md, line 79 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek2aTFCeNOs0FHFB4h:-Kek2aTGShQGYEa8E8fC:b-4047a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L79)):*
> ```Markdown
> ### Value encoding
> 
> KV values start with a four-byte checksum covering the whole KV and a
> ```

"the whole KV" - what's that? I think you may have meant "the whole KV pair".

---

*[docs/encoding.md, line 85 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek2jkjJ7zbQygaSn7I:-Kek2jkk8HXOJ80JdZ82:b-1pft9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L85)):*
> ```Markdown
> [pkg/roachpb/data.go](https://github.com/cockroachdb/cockroach/blob/master/pkg/roachpb/data.go)).
> For most column families, the value type is `TUPLE`, which indicates the
> following encoding. For each column in the column family sorted by ID,
> ```

"sorted by column ID"

---

*[docs/encoding.md, line 128 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek3cSpZBloCBxM_vfI:-Kek3cSpZBloCBxM_vfJ:bqxgwf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L128)):*
> ```Markdown
> followed by the usual encoding starting with the table ID but without
> the columns that appear in the parent table. The sentinel informs the
> decoder that the row does not belong to the parent table.
> ```

Please add an example here.

---

*[docs/encoding.md, line 192 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek4OH1jy4cx4x61qR7:-Kek4OH2y-_kTxAC0Qnz:b-91rliu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L192)):*
> ```Markdown
> columns, every column whose key encoding does not suffice for decoding
> needs to be encoded again. The encoded bytes appear at the end of the
> `BYTES` value (indexed columns followed by extra columns).
> ```

Can you provide an example here.

---

*[docs/encoding.md, line 199 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek4QN4BCPa5dOR_1Xs:-Kek4QN4BCPa5dOR_1Xt:b-8mlndb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L199)):*
> ```Markdown
> columns of the table must be a prefix (not necessarily proper) of the
> list of indexed columns. The encoding changes are the same as for
> table-table interleaving.
> ```

Please provide an example here too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Kek0Qu2zhOns4kMsnW8:b-mjegt8)*
<!-- Sent from Reviewable.io -->

knz commented:
This document is absolutely gorgeous and made me realize I had wanted to read it for a long time. Extremely well done.

---

Review status: 22 of 23 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Kek4XiuU8L8dgo4W0uS:b-ryw056)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Oof, sorry about the delay! I read the encoding doc on Thursday and it seemed correct. Wanted to do a deeper second read on Friday but forgot, I'll definitely get to this today.

FWIW, this doc seems like it wants to be a separate PR if you don't want to block this one on getting the doc LGTM'd

---

Review status: 21 of 23 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Kf7TZnsspDydyp_bLDJ:b-b3ty7d)*
<!-- Sent from Reviewable.io -->

danhhz commented:
encoding doc looks correct to me

---

Review status: 21 of 23 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[docs/encoding.md, line 9 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek0XNM3SNUovRZqg9u:-Kf8N-ZIQ_4lw76gHrS5:b-250ram) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L9)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

There needs to be a (short) explanation here that this encoding tends to change over time, that we have "format versions" that reflect these changes, and that we try to keep it backward-compatible.

Also mention here which specific format version this doc is explaining. (I think 3? @danhhz please confirm)
</blockquote></details>

= knz. an overview of how the format versions are managed and a high level of how they interact with data written under a previous versions would be nice

---

*[docs/encoding.md, line 12 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8MaLaVf5a3kmP6ZgK:-Kf8MaLbeMm4Oo4DJnsW:b-vdvlr) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L12)):*
> ```Markdown
> The Cockroach Labs blog post [SQL in CockroachDB: Mapping Table Data to
> Key-Value
> Storage](https://www.cockroachlabs.com/blog/sql-in-cockroachdb-mapping-table-data-to-key-value-storage/)
> ```

optional: if you're not aware of it, there is a "reference style" of link that I find makes these long ones more readable: https://daringfireball.net/projects/markdown/syntax#link

---

*[docs/encoding.md, line 13 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8MU75_HarVav9izw5:-Kf8MU75_HarVav9izw6:b-gd7ev) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L13)):*
> ```Markdown
> Key-Value
> Storage](https://www.cockroachlabs.com/blog/sql-in-cockroachdb-mapping-table-data-to-key-value-storage/)
> covers the first format version, which predates column families,
> ```

there's a blog for columns families, too (https://www.cockroachlabs.com/blog/sql-cockroachdb-column-families/)

maybe also worth including links to the rfcs

---

*[docs/encoding.md, line 29 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8NNwN5v7LjUz9fbg9:-Kf8NNwN5v7LjUz9fbgA:bxgkioj) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L29)):*
> ```Markdown
> family is a maximal subset of columns with the same column family ID.
> For more details, see
> [pkg/sql/sqlbase/structured.proto](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/structured.proto).
> ```

maybe pick some beta tag for these links instead of master, so they don't break if someone renames the file. this would also let you link to the line numbers if you like

---

*[docs/encoding.md, line 129 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8OaQxzuV7QpToEzuY:-Kf8OaQyLlISA4XMdsmt:b-4hcueb) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L129)):*
> ```Markdown
> Unlike the `TUPLE` encoding, this encoding did not need to be
> prefix-free, which was a boon for strings. CRDB still constructs
> non-sentinel KV values this way for single-column column families when
> ```

mention that this is necessary for backward compatibility, plz

---

*[docs/encoding.md, line 236 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8P5tRSyTtnGQ8CsTG:-Kf8P5tRSyTtnGQ8CsTH:b-p3mm28) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L236)):*
> ```Markdown
> belongs to the column family with ID 0).
> 
> ### Interleaving
> ```

I find explanations of interleaving to be more clear if they're right from the start about interleaving any index into any other index. in this case using "table" to mean "primary index data" makes me feel its harder to reason about. ymmv

---

*[docs/encoding.md, line 245 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8PYmkvhoc9DWRkUWl:-Kf8PYmkvhoc9DWRkUWm:bq0r787) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L245)):*
> ```Markdown
> necessarily proper) of the other table, the child. To encode a KV key
> for a row in the child table, encode it as if it were a row in the
> parent table, then append the interleaving sentinel
> ```

mention that the interleaving sentinel appears in the column family slot

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-Kf8TO43lCa5b2AE1Eks:b-qgi41h)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 1 files at r14, 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/encoding.md, line 146 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8ZIuCaIFYWKC3hYr2:-Kf8ZIuCaIFYWKC3hYr3:buikv7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L146)):*
> ```Markdown
> dump follows.
> 
>     DROP DATABASE IF EXISTS bank;
> ```

Enclose the code/data examples with the delimiter "```" to keep them preformatted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Revised the doc per comments below. As I was creating the examples, I also noticed a bug in how `CompositeColumnIDs` was computed -- apparently IDs can be assigned multiple times, and the field was not cleared before the appends. Fixed that.

---

Review status: 20 of 23 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/encoding.md, line 9 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek0XNM3SNUovRZqg9u:-Kf7Ndq-p8eFZxqKoxdc:b4ay2h3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L9)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

= knz. an overview of how the format versions are managed and a high level of how they interact with data written under a previous versions would be nice
</blockquote></details>

Done. I wrote something on versioning in the single-column column family section, but did you want me to say more?

---

*[docs/encoding.md, line 40 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek1DYORm2hoUC7WIh8:-KfBwt0kGDWfDv_Nrotp:b-un9o4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L40)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The word "sentinel key" must be mentioned and defined here.

Please provide three example KV pairs here: one with no null values, one with null values on only some columns of a family, and one with null values on all columns but that is a PK (= example sentinel key)
</blockquote></details>

The first example has this property.

---

*[docs/encoding.md, line 50 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek1uXVo6t0qRZnsG96:-Kf7NklwHG-tuB9M6eoJ:b-jbknl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L50)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"The encoding of individual fields is chosen so that it is always possible to decode a concatenated key into its original components."

Also you need to define the term "field code word".
</blockquote></details>

Done. I rewrote to not use that technical term.

---

*[docs/encoding.md, line 60 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek2QQ9iweujRwyWsze:-Kf823U9exjr4A9wW8rs:b4re2gt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L60)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think it would be useful to provide a table with some of the most commonly used SQL types and their code word, as examples, here.
</blockquote></details>

This is more detail than I wanted to go into, especially given that `cockroach debug rocksdb scan` does field decoding.

Perhaps the worked example below will be enlightening.

---

*[docs/encoding.md, line 79 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek2aTFCeNOs0FHFB4h:-Keng6jU2GyG48abyxzH:bwzzjjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L79)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"the whole KV" - what's that? I think you may have meant "the whole KV pair".
</blockquote></details>

Defined KV above as key-value pair above, but since this is confusing, I changed it to key-value and changed all bare instances of KV to KV pair.

---

*[docs/encoding.md, line 85 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek2jkjJ7zbQygaSn7I:-KengLRwPBpIlqtHqGPa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L85)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"sorted by column ID"
</blockquote></details>

Done.

---

*[docs/encoding.md, line 128 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek3cSpZBloCBxM_vfI:-KfCB2LkV2EvxZFr17Wl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L128)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please add an example here.
</blockquote></details>

Done.

---

*[docs/encoding.md, line 192 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek4OH1jy4cx4x61qR7:-KfCB2yCZ4bplq0BULGx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L192)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you provide an example here.
</blockquote></details>

Done.

---

*[docs/encoding.md, line 199 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kek4QN4BCPa5dOR_1Xs:-KfCB3D75JI_8kmazKre:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed884a8b7389f0c2e4eb6fac14e5e4ee7c338ecc/docs/encoding.md#L199)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please provide an example here too.
</blockquote></details>

Done.

---

*[docs/encoding.md, line 12 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8MaLaVf5a3kmP6ZgK:-KfBlLZuR50IXWKbH7iY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L12)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

optional: if you're not aware of it, there is a "reference style" of link that I find makes these long ones more readable: https://daringfireball.net/projects/markdown/syntax#link
</blockquote></details>

Done.

---

*[docs/encoding.md, line 13 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8MU75_HarVav9izw5:-KfBx2q4qhXsHWCTlShi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L13)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

there's a blog for columns families, too (https://www.cockroachlabs.com/blog/sql-cockroachdb-column-families/)

maybe also worth including links to the rfcs
</blockquote></details>

Done.

---

*[docs/encoding.md, line 29 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8NNwN5v7LjUz9fbg9:-KfCANN6VZ5ppjLROkmM:b-csoxho) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L29)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

maybe pick some beta tag for these links instead of master, so they don't break if someone renames the file. this would also let you link to the line numbers if you like
</blockquote></details>

The downside is that data types added after the beta won't show. Since the filenames have been pretty stable, I'm inclined to pass on this suggestion.

---

*[docs/encoding.md, line 129 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8OaQxzuV7QpToEzuY:-KfCAo44ut_UfcgrQhZB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L129)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

mention that this is necessary for backward compatibility, plz
</blockquote></details>

Done.

---

*[docs/encoding.md, line 146 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8ZIuCaIFYWKC3hYr2:-KfBvvFHt839yj80nTot:bjm131n) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L146)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Enclose the code/data examples with the delimiter "```" to keep them preformatted.
</blockquote></details>

4-space indent was the original Markdown code block syntax. I'd use backticks, but Pandoc's Markdown writer (the tool I use to keep this document tidy) doesn't support them.

---

*[docs/encoding.md, line 236 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8P5tRSyTtnGQ8CsTG:-KfCAq03eGGWM4LlyK17:b-7y334g) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L236)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I find explanations of interleaving to be more clear if they're right from the start about interleaving any index into any other index. in this case using "table" to mean "primary index data" makes me feel its harder to reason about. ymmv
</blockquote></details>

Moved this to its own section.

---

*[docs/encoding.md, line 245 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8PYmkvhoc9DWRkUWl:-KfBwkou7GwArYgGNMr1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L245)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

mention that the interleaving sentinel appears in the column family slot
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-KfCtAubIs3YWdZtHswk:btpmr9p)*
<!-- Sent from Reviewable.io -->

knz commented:
Well spotted.

I'm happy with the result. LGTM

---

Reviewed 2 of 3 files at r16, 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[docs/encoding.md, line 146 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-Kf8ZIuCaIFYWKC3hYr2:-KfD3HFJgXNlgbCSkViS:bihkiuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf93b8ea0c4d18cd80e76bb7011e385a24a59ed0/docs/encoding.md#L146)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

4-space indent was the original Markdown code block syntax. I'd use backticks, but Pandoc's Markdown writer (the tool I use to keep this document tidy) doesn't support them.
</blockquote></details>

Oh nice, TIL

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-KfD3XkeBJRajW1ypOCG:b3kb4oh)*
<!-- Sent from Reviewable.io -->

danhhz commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> This is great stuff! Thanks for doing it!

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[docs/encoding.md, line 124 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KfHPAclpzgxm13rhz53:-KfHPAcmI1L8rRQbXi9R:bulbner) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cdaefe0f763b54bba14ce03cf2bca7e14ae705/docs/encoding.md#L124)):*
> ```Markdown
> did not need to be prefix-free, which was a boon for strings.
> 
> On upgrading to format version 2 or higher, CRDB puts each column in a
> ```

puts each column _in existing tables_

---

*[docs/encoding.md, line 126 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KfHPHoH8gC6rGtYQf1K:-KfHPHoH8gC6rGtYQf1L:b-ayoeuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cdaefe0f763b54bba14ce03cf2bca7e14ae705/docs/encoding.md#L126)):*
> ```Markdown
> On upgrading to format version 2 or higher, CRDB puts each column in a
> column family whose ID is the same as the column ID. This allows
> backward-compatible decoding. To save space, the encoder in fact still
> ```

this is also necessary for backward compatibility, not just to save space

---

*[docs/encoding.md, line 265 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KfHPtukv52VMczPnabe:-KfHPtulfKf8palt0MQZ:b603kvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cdaefe0f763b54bba14ce03cf2bca7e14ae705/docs/encoding.md#L265)):*
> ```Markdown
> Users also specify whether a secondary index should be unique. Unique
> secondary indexes constrain the table data not to have two rows that
> agree on the indexed columns.
> ```

nit: "agree" feels a little ambiguous here

---

*[docs/encoding.md, line 413 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KfHQcp2JsSKlow-GxN1:-KfHQcp37FosYr_8X3it:b5gjbja) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cdaefe0f763b54bba14ce03cf2bca7e14ae705/docs/encoding.md#L413)):*
> ```Markdown
> 
> Note that the parent may itself be interleaved. In general, the
> interleaving relationships constitute an [arborescence].
> ```

til: arborescence

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-:-KfHQmyny1-8BxQvUHJP:b-lint8o)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Review status: 21 of 23 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[docs/encoding.md, line 124 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KfHPAclpzgxm13rhz53:-KfHTZ5smMpyLxkQB6oh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cdaefe0f763b54bba14ce03cf2bca7e14ae705/docs/encoding.md#L124)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

puts each column _in existing tables_
</blockquote></details>

Done.

---

*[docs/encoding.md, line 126 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KfHPHoH8gC6rGtYQf1K:-KfHTZTmLCh-Sy8QW22L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cdaefe0f763b54bba14ce03cf2bca7e14ae705/docs/encoding.md#L126)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this is also necessary for backward compatibility, not just to save space
</blockquote></details>

Done.

---

*[docs/encoding.md, line 265 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609#-KfHPtukv52VMczPnabe:-KfHT_fbKlcTr0Nmh3QD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4cdaefe0f763b54bba14ce03cf2bca7e14ae705/docs/encoding.md#L265)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: "agree" feels a little ambiguous here
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13609)*
<!-- Sent from Reviewable.io -->
